// drivewealth-back-office-javascript/src/types/Advisors.js
// @flow

export type ProductDetails = {
    name: string,
    description: string,
    clientProductID: string,
    status: "ACTIVE" | "INACTIVE",
    metrics: {
        expectedReturn: number,
        expectedRisk: number,
        maxYearlyLoss: number,
        annualTurnover: number,
    },
};

export type EditProduct = {
    name?: string,
    description?: string,
    clientProductID?: string,
    status?: "ACTIVE" | "INACTIVE",
    metrics?: {
        expectedReturn: number,
        expectedRisk: number,
        maxYearlyLoss: number,
        annualTurnover: number,
    },
}

export type Allocation = {
    accountID: string,
    accountNo: string,
    qty: number,
    comment: string,
};

export type Order = {
    orderID: string,
    allocations: Array<Allocation>,
};

export type Holding = {
    type: string,
    id: string,
    target: number,
    instrumentID: string,
    symbol?: string,
    name?: string,
};

export type Trigger = {
    type: string,
    maxAllowed: string,
    child?: string,
    lowerBound: number,
    upperBound: number,
};

export type AllocationBody = {
    userID: string,
    clientListID: string,
    note: string,
    orders: Array<Order>,
};

export type Portfolio = {
    userID: string,
    name: string,
    clientPortfolioID: string,
    description: string,
    holdings: Array<Holding>,
    triggers: Array<Trigger>,
};

export type Fund = {
    userID: string,
    name: string,
    clientFundID: string,
    description: string,
    holdings: Array<Holding>,
    triggers: Array<Trigger>,
}
