// drivewealth-back-office-javascript/src/types/Document.js
// @flow
import type { UserDetails } from "./User";

/**
 * @memberof Types
 */
export type Document = {
	documentID: string,
	type: {
		name: string,
		description: string,
	},
	ipAddress: string,
	receivedVia: string,
	receivedWhen: string,
	status: {
		name: string,
		description: string,
	},
};

/**
 * @memberof Types
 */
export type PendingCount = {
	language: {
		languageID: string,
		locale: string,
		active: boolean,
		nameEnglish: string,
		nameNative: string,
	},
	userCount: number,
	documentCount: number,
};

/**
 * @memberof Types
 */
export type PendingDocument = {
	documentID: string,
	documentType: string,
	active: boolean,
	status: number,
	imageWidth: number,
	imageHeight: number,
	approvedWhen: string,
	approvedBy: string,
	receivedWhen: string,
	receivedVia: string,
};

/**
 * @memberof Types
 */
export type PendingDocuments = {
	user: UserDetails,
	documentLanguageID: string,
	documents: Array<PendingDocument>,
	ttl: number,
};

/**
 * @memberof Types
 */
export type ReviewedDocument = {
	documentID: string,
	documentType: string,
	status: string,
	rejectionReason: void | { type: string },
};
