// drivewealth-back-office-javascript/src/types/Instrument.js
// @flow
/**
 * @memberof Types
 */
export type Instrument = {
	symbol: string,
	reutersPrimaryRic: string,
	name: string,
	description: string,
	sector: null,
	longOnly: boolean,
	orderSizeMax: number,
	orderSizeMin: number,
	orderSizeStep: number,
	exchangeNickelSpread: boolean,
	close: number,
	descriptionChinese: string | null,
	id: string,
	type: string,
	exchange: string,
	url: string,
	status: string,
	exchangeThresholdOverride: number,
	houseThresholdOverride: null,
	closePrior: number,
	image: string,
};

/**
 * @memberof Types
 */
export type InstrumentCreateDetails = {
	symbol: string,
	reutersPrimaryRic: string,
	name: string,
	status: string,
	description: string,
	url: string,
	image: string,
	exchange: string,
	type: string,
	longOnly: boolean,
	orderSizeMax: number,
	orderSizeMin: number,
	orderSizeStep: number,
	exchangeNickelSpread: boolean,
}

/**
 * @memberof Types
 */
export type InstrumentBasic = {
	symbol: string,
	name: string,
	id: string,
	status: string
}

/**
 * @memberof Types
 */
export type CorporateActionEventTypes = {
	IPO_PRICINGS: string,
	OTHER_BROKERAGES: string,
	ECONOMIC_EVENTS: string,
	OTHER_CORPORATE: string,
	CORPORATE_CALLS_AND_PRESENTATIONS: string,
	IPO_LOCKUP_EXPIRATIONS: string,
	IPO_WITHDRAWALS: string,
	CORPORATE_ANALYST_MEETING: string,
	EARNINGS_CALLS_AND_PRESENTATIONS: string,
	STOCK_SPLITS: string,
	SECONDARY_WITHDRAWALS: string,
	MERGER_AND_ACQUISTIONS_CALLS_AND_PRESENTATIONS: string,
	BROKERAGE_ANALYST_MEETING: string,
	GUIDANCE_CALLS_AND_PRESENTATIONS: string,
	BROKERAGE_ANALYST_CALLS: string,
	SALES_AND_TRADING_STATEMENT_CALLS_RELEASES: string,
	SECONDARY_FILINGS: string,
	SYNDICATE_ROADSHOWS: string,
	CONFERENCES: string,
	GENERAL_AND_POLITICAL_EVENTS: string,
	COMPANY_VISITS: string,
	EX_DIVIDENDS: string,
	SALES_AND_TRADING_STATEMENT_CALLS_AND_PRESENTATIONS: string,
	SHAREHOLDER_AND_ANNUAL_MEETINGS: string,
	CONFERENCE_PRESENTATIONS: string,
	EARNINGS_RELEASES: string,
	IPO_FILINGS: string,
	SECONDARY_PRICINGS: string,
};
