// drivewealth-back-office-javascript/src/types/MoneyMovement.js
// @flow
/**
 * @memberof Types
 */
export type MoneyMovementSummary = {
	accountDetails: {
		accountID: string,
		accountManagementType: {
			name: string,
			description: string,
		},
		accountNo: string,
		accountType: {
			name: string,
			description: string,
		},
	},
	amount: number,
	currency: {
		description: string,
		name: string,
		symbol: string,
	},
	fees: Array<{
		description: string,
		amount: number,
		wlpFinTranTypeID: string,
		currency: string,
		symbol: string,
	}>,
	id: string,
	status: {
		id: number,
		message: string,
		updated: string,
	},
	type: string,
	userDetails: {
		email: string,
		firstName: string,
		lastName: string,
		userID: string,
		username: string,
		wlpID: string,
	},
};

/**
 * @memberof Types
 */
export type MoneyMovementDetails = {
	accountDetails: {
		accountID: string,
		accountManagementType: {
			name: string,
			description: string,
		},
		accountNo: string,
		accountType: {
			name: string,
			description: string,
		},
	},
	amount: number,
	batch: boolean,
	category: string,
	created: string,
	currency: {
		name: string,
		description: string,
		symbol: string,
	},
	details: {
		country: string,
		bankAccountType: string,
		bankRoutingNumber: string,
		bankAccountNumber: string,
		accountHolderName: string,
		personalID: string,
		phone: string,
		email: string,
	},
	fees: Array<{
		description: string,
		amount: number,
		wlpFinTranTypeID: string,
		currency: string,
		symbol: string,
	}>,
	id: string,
	note: string,
	status: {
		id: number,
		message: string,
		updated: string,
	},
	statusHistory: Array<{
		status: number,
		statusMessage: string,
		created: string,
		updatedBy: string,
		comment: string,
	}>,
	type: string,
	userDetails: {
		email: string,
		firstName: string,
		lastName: string,
		parentIB: {
			id: string,
			name: string,
		},
		userID: string,
		username: string,
		wlpID: string,
	},
	wlpFinTranTypeID: string,
};

export type PatchBody = {
	bankAccountID?: string,
	amount?: string,
	frequency?: string,
	active?: boolean,
};
