// drivewealth-back-office-javascript/src/types/Order.js
// @flow
import Order from "../modules/Order";

/**
 * @memberof Types
 */
export type OrderDetails = {
	orderID: string,
	orderNo: string,
	side: string,
	symbol: string,
	createdWhen: string,
	isoTimeRestingOrderExpires: string | null,
	orderStatus: string,
	orderType: string,
	orderCashAmt: number,
	orderQty: number,
	cumQty: number,
	limitPrice: number,
	stopPrice: number,
};

/**
 * @memberof Types
 */
export type OrderCreateDetails = {
	orderType: $Keys<typeof Order.TYPES>,
	symbol: string,
	side: $Keys<typeof Order.SIDES>,
	quantity?: number,
	amountCash?: number,
	price?: number,
	commission?: number,
	orderNotes?: string,
	clientNotes?: string,
	marginLiquidate?: boolean,
};

/**
 * @memberof Types
 */
export type OrderCreateResponse = {
	orderID: string,
	orderNo: string,
	symbol: string,
	cumQty: number,
	orderStatus: string,
	orderType: string,
	orderQty: number,
	limitPrice: number,
	stopPrice: number,
	executedPrice: number,
	side: string,
	createdWhen: string,
	updatedWhen: string,
	updatedReason: string,
	commission: number,
	commissionDesc: string,
	isoTimeRestingOrderExpires: string,
	executedWhen: string,
	realizedPL: number | null,
	orderCashAmt: number,
};
