// drivewealth-back-office-javascript/src/types/Reports.js
// @flow
/**
 * @memberof Types
 */
export type ReportViolations = {
	accountNo: string,
	currViolations: number,
	prevViolations: number,
	restrictionStatus: string,
};

/**
 * @memberof Types
 */
export type ReportTicket = {
	accountID: string,
	accountNo: string,
	accountType: number,
	commission: number,
	cumQty: number,
	createdWhen: string,
	execBroker: string,
	execID: string,
	execType: string,
	lastPx: number,
	lastShares: number,
	leavesQty: number,
	orderID: string,
	orderNo: string,
	orderQty: number,
	ordStatus: string,
	side: string,
	symbol: string,
	transactTime: string,
	routedVia: string,
	fixLimitPrice: number,
	fixSystemOrderID: number,
	stopPx: number,
	mitTriggerPrice: number,
	ordType: string,
	rateAsk: number,
	rateBid: number,
	internalMemo: string,
	dnb: boolean,
	doNotSendToIcbc: boolean,
	amountCash: number,
	pnl: number,
	lastMkt: string,
	avgPx: number,
	orderCreationDate: string,
	instrumentID: string,
	feeSec: number,
	feeTaf: number,
	feeBase: number,
	feeXtraShares: number,
	feeExchange: number,
	reportingSystem: string,
	commissionOverride: {
		commissionOverrideType: string,
		commissionOverrideDesc: string,
	},
	journalExcessFunds: boolean,
	violationExempt: boolean,
	backOfficeShareAdjustment: boolean,
};

/**
 * @memberof Types
 * @description Optional properties for filtering a report.
 */
export type ExtraFields = {
	DateStart: ?string,
	DateEnd: ?string,
	IbUserName: ?string,
	AccountNumber: ?string,
	AccountType: ?string,
	Date: ?string,
	TradingType: ?string,
	subscription: ?boolean,
};

/**
 * @memberof Types
 */
export type ReportFile = {
	displayName: string,
	fileKey: string,
}
