// drivewealth-back-office-javascript/src/types/Transaction.js
// @flow
/**
 * @memberof Types
 */
export type Transaction = {
	orderId: string,
	orderNo: string,
	symbol: string,
	side: string,
	orderStatus: string,
	orderType: string,
	orderQty: number,
	cumQty: number,
	limitPrice: number,
	stopPrice: number,
	executedPrice: number,
	createdWhen: string,
	executedWhen: string,
	updatedWhen: string,
	updatedReason: string,
	isoTimeRestingOrderExpires: string,
	commission: number,
	commissionDesc: string,
	realizedPL: null,
	orderCashAmt: number,
};

/**
 * @memberof Types
 */
export type TransactionType = {
	active: boolean,
	description: string,
	finTranTypeID: string,
	wlpFinTranTypeID: string,
	moneyTypeID: string,
	currency: string,
	clearingAccount: string,
	glCode: string,
	glSubCode: string,
};

/**
 * @memberof Types
 */
export type TransactionTypeList = {
	wlpID: string,
	wlpName: string,
	types: Array<TransactionType>,
};

/**
 * @memberof Types
 */
export type TransactionCreateResponse = {
	accountID: string,
	accountNo: string,
	accountAmount: number,
	accountBalance: number,
	comment: string,
	currencyID: string,
	ipAddress: string,
	finTranID: string,
	finTranTypeID: string,
	tranAmount: number,
	tranSource: string,
	tranWhen: string,
	wlpFinTranTypeID: string,
	accountType: number,
	dnb: boolean,
	sendCommissionToInteliclear: boolean,
	feeSec: number,
	feeTaf: number,
	feeBase: number,
	feeXtraShares: number,
	feeExchange: number,
};

/**
 * @memberof Types
 */
export type TransactionDetails = {
	accountAmount: number,
	accountBalance: number,
	accountType: string,
	comment: string,
	dnb: boolean,
	finTranID: string,
	finTranTypeID: string,
	feeSec: number,
	feeTaf: number,
	feeBase: number,
	feeXtraShares: number,
	feeExchange: number,
	fillQty: number,
	fillPx: number,
	sendCommissionToInteliclear: boolean,
	systemAmount: number,
	tranAmount: number,
	tranSource: string,
	tranWhen: string,
	wlpAmount: number,
};

/**
 * @memberof Types
 */
export type TransactionBatchResponse = {
	accountNo: string,
	transactionType: string,
	batchCode: string,
	amount: number,
	status: string,
	comment: string,
	correlationID: string,
	finTran: {
		accountID: string,
		accountNo: string,
		accountAmount: number,
		accountBalance: number,
		comment: string,
		currencyID: string,
		ipAddress: string,
		finTranID: string,
		finTranTypeID: string,
		tranAmount: number,
		tranSource: string,
		tranWhen: string,
		wlpFinTranTypeID: string,
		accountType: number,
		dnb: boolean,
		sendCommissionToInteliclear: boolean,
		feeSec: number,
		feeTaf: number,
		feeBase: number,
		feeXtraShares: number,
		feeExchange: number,
	},
};

export type PaginationOptions = {|
	limit: number,
|} | {|
	limit: ? number,
	offset: string,
	direction: "next" | "prev",
|};
