// drivewealth-back-office-javascript/src/modules/Roles.js
// @flow
import request from "./request";
import type { UserDetails } from "../types/User";
import type {
	Role,
	RoleGroup,
	PermissionRole,
} from "../types/Roles";

/**
 * @module
 * @description Roles allow clients to send REST requests to the Permissions API.
 * @example import { Permissions } from "drivewealth-back-office-javascript";
 */
export default class Roles {
	/**
	 * @static
	 */
	static getAllPermissionGroups(): Promise<Array<RoleGroup>> {
		return request({
			endpoint: `/permissions/groups`,
		}).then(data => data.body);
	};

	/**
	 * @static
	 */
	static getAllPermissions(): Promise<Array<Role>> {
		return request({
			endpoint: `/permissions`,
		}).then(data => data.body);
	};

	/**
	 * @static
	 */
	static createPermissionGroup(
		active: boolean,
		description: string,
		permissionSet: Array<string>,
		role: string,
	): Promise<RoleGroup> {
		const body = {
			active,
			description,
			permissionSet,
			role,
		};
		return request({
			method: "POST",
			endpoint: `/permissions/groups`,
			body,
		}).then(data => data.body);
	};

	/**
	 * @static
	 */
	static editRole(
		id: string,
		role: RoleGroup,
	): Promise<PermissionRole> {
		const body = role;
		return request ({
			method: "PATCH",
			endpoint: `/permissions/groups/${id}`,
			body,
		}).then(data => data.body);
	};
};
