// drivewealth-back-office-javascript/src/modules/Violation.js
// @flow
import request from "./request";
import type { ViolationCreateDetails } from "../types/Violation";

/**
 * @module
 * @example import { Violation } from "drivewealth-back-office-javascript";
 */
export default class Violation {
	/**
	 * Create a new violation tied to an account.
	 */
	static create(accountID: string, details: ViolationCreateDetails): Promise<void> {
		return request({
			method: "POST",
			endpoint: `/accounts/${accountID}/violations`,
			body: details,
		}).then(() => undefined);
	}

	/**
	 * Get violations of an account.
	 */
	static get(accountID: string): Promise<{
		goodFaithViolations: Array<Violation>,
		patternDayTrades: number | null,
		legacyGoodFaithViolations: number,
		message: string,
	}> {
		return request({ endpoint: `/accounts/${accountID}/violations` }).then(data => data.body);
	}

	/**
	 * Delete a violation from an account.
	 */
	static delete(accountID: string, violationID: string): Promise<void> {
		return request({
			method: "DELETE",
			endpoint: `/accounts/${accountID}/violations/${violationID}`,
		}).then(() => undefined);
	}
}
